substrate-archive/src/sql_block_builder.rs

```diff
@@ -23,85 +23,55 @@ use std::marker::PhantomData;
 use codec::{Decode, Encode};
 use sp_runtime::{
 	generic::SignedBlock,
-	traits::{Block as BlockT, DigestFor, Header as HeaderT},
+	traits::{Block as BlockT, Header as HeaderT},
 };
 
 use substrate_archive_common::{models::BlockModel, types, Error};
 
-struct GenericParts<B: BlockT>(
-	<B::Header as HeaderT>::Hash,
-	<B::Header as HeaderT>::Hash,
-	<B::Header as HeaderT>::Hash,
-);
-
-pub struct BlockBuilder<B: BlockT> {
+pub struct SqlBlockBuilder<B: BlockT> {
 	_marker: PhantomData<B>,
 }
 
-impl<'a, B: BlockT> BlockBuilder<B> {
+impl<'a, B: BlockT> SqlBlockBuilder<B> {
 	/// With a vector of BlockModel
 	pub fn with_vec(blocks: Vec<BlockModel>) -> Result<Vec<types::Block<B>>, Error> {
 		blocks
 			.into_iter()
 			.map(|b| {
-				let (b, s) = Self::with_single(b)?;
-				let b = SignedBlock { block: b, justification: None };
-				Ok(types::Block::new(b, s))
+				let (block, spec) = Self::with_single(b)?;
+				let block = SignedBlock { block, justification: None };
+				Ok(types::Block::new(block, spec))
 			})
 			.collect()
 	}
 
+	/// With a single BlockModel
 	pub fn with_single(block: BlockModel) -> Result<(B, u32), Error> {
-		let digest: DigestFor<B> = Decode::decode(&mut block.digest.as_slice())?;
-		let GenericParts(parent_hash, state_root, extrinsics_root) = Self::generic_parts_from(
-			block.parent_hash.as_slice(),
-			block.state_root.as_slice(),
-			block.extrinsics_root.as_slice(),
-		)?;
-		let num: <B::Header as HeaderT>::Number = Decode::decode(&mut (block.block_num as u32).encode().as_slice())?;
-
-		let header = <B::Header as HeaderT>::new(num, extrinsics_root, state_root, parent_hash, digest);
-		let ext: Vec<B::Extrinsic> = Decode::decode(&mut block.ext.as_slice())?;
-		let spec = block.spec;
-		Ok((B::new(header, ext), spec as u32))
+		let BlockDecoder { header, ext, spec } = BlockDecoder::<B>::decode(block)?;
+		let block = B::new(header, ext);
+		Ok((block, spec))
 	}
+}
 
-	fn generic_parts_from(
-		mut parent_hash: &[u8],
-		mut state_root: &[u8],
-		mut extrinsics_root: &[u8],
-	) -> Result<GenericParts<B>, Error> {
-		Ok(GenericParts(
-			Decode::decode(&mut parent_hash)?,
-			Decode::decode(&mut state_root)?,
-			Decode::decode(&mut extrinsics_root)?,
-		))
-	}
+struct BlockDecoder<B: BlockT> {
+	header: B::Header,
+	ext: Vec<B::Extrinsic>,
+	spec: u32,
```

Comment:
what's the reason why `BlockModel` uses i32 as spec and `BlockDecoder` uses u32?

Reply:
`BlockMode`l is the result of a query from Postgres, and since Postgres does not support unsigned integers, only the signed i32/i64 rust types implement `Encode`/`Decode` in SQLx.